plugin/kotel/tracing.go

```diff
@@ -0,0 +1,162 @@
+	)
+	return t
+}
+
+// TracerProvider takes a trace.TracerProvider and applies it to the Tracer
+// If none is specified, the global provider is used.
+func TracerProvider(provider trace.TracerProvider) TracingOption {
+	return tracingOptionFunc(func(t *Tracer) {
+		if t != nil {
+			t.tracerProvider = provider
+		}
+	})
+}
+
+// TracerPropagator takes a propagation.TextMapPropagator and applies it to the Tracer
+// If none is specified, the global Propagator is used.
+func TracerPropagator(propagator propagation.TextMapPropagator) TracingOption {
+	return tracingOptionFunc(func(t *Tracer) {
+		if t != nil {
+			t.propagators = propagator
+		}
+	})
+}
+
+// Hooks ---------------------------------------------------------------------
+
+// OnProduceRecordBuffered Begins the producer span
+func (t *Tracer) OnProduceRecordBuffered(r *kgo.Record) {
+	attrs := []attribute.KeyValue{
+		semconv.MessagingSystemKey.String("kafka"),
+		semconv.MessagingDestinationKindTopic,
+		semconv.MessagingDestinationKey.String(r.Topic),
+		semconv.MessagingKafkaClientIDKey.String(strconv.FormatInt(r.ProducerID, 10)),
+		semconv.MessagingOperationKey.String("send"),
+	}
+
+	opts := []trace.SpanStartOption{
+		trace.WithAttributes(attrs...),
+		trace.WithSpanKind(trace.SpanKindProducer),
+	}
+
+	spanContext, _ := t.tracer.Start(r.Context, fmt.Sprintf("%s send", r.Topic), opts...)
+
+	t.propagators.Inject(spanContext, NewRecordCarrier(r))
+
+	r.Context = spanContext
+}
+
+// OnProduceRecordUnbuffered Completes the producer span
+func (t *Tracer) OnProduceRecordUnbuffered(r *kgo.Record, err error) {
+	span := trace.SpanFromContext(r.Context)
+
+	span.SetAttributes(
+		semconv.MessagingMessageIDKey.String(strconv.FormatInt(r.Offset, 10)),
+		semconv.MessagingKafkaPartitionKey.Int64(int64(r.Partition)),
+	)
+
+	if err != nil {
+		span.SetStatus(codes.Error, err.Error())
+		span.RecordError(err)
+	}
+
+	span.End()
+}
+
+// OnFetchRecordBuffered Starts and completes the consumer spans
+func (t *Tracer) OnFetchRecordBuffered(r *kgo.Record) {
+	if r.Context == nil {
+		r.Context = context.Background()
+	}
+	ctx := r.Context
+	ctx = t.propagators.Extract(ctx, NewRecordCarrier(r))
+
+	attrs := []attribute.KeyValue{
+		semconv.MessagingSystemKey.String("kafka"),
+		semconv.MessagingDestinationKindTopic,
+		semconv.MessagingDestinationKey.String(r.Topic),
+		semconv.MessagingOperationReceive,
+		semconv.MessagingMessageIDKey.String(strconv.FormatInt(r.Offset, 10)),
+		semconv.MessagingKafkaPartitionKey.Int64(int64(r.Partition)),
+	}
+
+	var nextSpanContext = ctx
+	var nextSpan trace.Span
+
+	// Optionally add the LogAppend span if the record timestamp is set in Kafka
+	if r.Attrs.TimestampType() == 1 {
+		opts := []trace.SpanStartOption{
+			trace.WithTimestamp(r.Timestamp),
+			trace.WithAttributes(attrs...),
+			trace.WithSpanKind(trace.SpanKindConsumer),
+		}
+
+		nextSpanContext, nextSpan = t.tracer.Start(ctx, fmt.Sprintf("%s logappend", r.Topic), opts...)
+		nextSpan.End()
```

Comment:
Should this fall into the below logic? I'm wondering why there are two spans here.

Reply:
As per a requirement, we need to track events both when they land in Kafka (conditional log append span) and when the message is received by the client (aka receive span). This will give us a complete picture of message flows, and a better understanding of how the system is working.

Essentially these two spans record different timestamps.  Here is a screenshot to visualize this

![Kafka tracing](https://user-images.githubusercontent.com/13387464/211967592-346a9c32-21c2-4725-9719-1028121d94b6.jpg)

I've brought up the log append span as a debatable aspect of this implementation in the PR introduction. @brunsgaard has more knowledge and experience on this, would be valuable to hear their thoughts on the significance of it, and why it is beneficial to have it as part of this implementation.

However, if it's not deemed important, I am open to the idea of removing the log append span, and potentially achieving a similar goal by passing the `r.Timestamp` to the receive span attributes, to track the time when the message was received. However, I don't know whether this would cause a cardinality explosion on the metrics database.

Please let me know your thoughts on this.

Reply:
I see, this makes sense -- I need to read the otel docs a bit to understand why these spans are being ended immediately (unless you have a quick tldr), but I see the point of tracking when they landed on the broker. +1 to keeping this.

Reply:
We want to end the receive span before starting the process span. Also, if we end the receive span in the hook (hidden away from users), we can guarantee that it will always be closed. Although, starting and ending the process span is up to the user. We are open to any suggestions.

Reply:
P.S., the latest commit has removed the log append span to simplify the code further.